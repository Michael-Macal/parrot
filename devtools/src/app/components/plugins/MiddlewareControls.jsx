import React, { Component } from 'react';
import PropTypes from 'prop-types';
import url from 'url';

import SelectField from 'material-ui/SelectField';
import MenuItem from 'material-ui/MenuItem';
import CircularProgress from 'material-ui/CircularProgress';

import fetchApi from '../../utils/fetchApi';

class MiddlewareControls extends Component {
  static propTypes = {
    url: PropTypes.string.isRequired, // eslint-disable-line react/no-unused-prop-types
  }

  state = {
    loading: true,
    scenarios: {},
    scenario: '',
    error: null,
  };

  async componentDidMount() {
    try {
      const resp = await fetchApi(this.props.url, '/parrot/scenarios');
      const scenarios = await resp.json();
      const keys = Object.keys(scenarios);

      this.setState({ // eslint-disable-line react/no-did-mount-set-state
        loading: false,
        scenarios,
        scenario: keys[0],
      });
    } catch (error) {
      this.setState({ error }); // eslint-disable-line react/no-did-mount-set-state
    }
  }

  setScenario = (event, key, scenario) => {
    this.setState({
      scenario,
    }, async () => {
      try {
        fetchApi(this.props.url, '/parrot/scenario', {
          method: 'POST',
          headers: {
            Accept: 'application/json',
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({
            scenario,
          }),
        });
      } catch ({ message: error }) {
        this.setState({ error });
      }
    });
  }

  createUrl = pathname => url.format({ ...url.parse(this.props.url), pathname })

  render() {
    if (this.state.error) {
      return <pre>{this.state.error}</pre>;
    } else if (this.state.loading) {
      return <CircularProgress />;
    }

    return (
      <SelectField
        style={{width: '100%'}}
        floatingLabelText="Selected Scenario"
        value={this.state.scenario}
        onChange={this.setScenario}
      >
        {Object.keys(this.state.scenarios).map(scenario =>
          <MenuItem key={scenario} value={scenario} primaryText={scenario} />,
        )}
      </SelectField>
    );
  }
}

export default MiddlewareControls;
