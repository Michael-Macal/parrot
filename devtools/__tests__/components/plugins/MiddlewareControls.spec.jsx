import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import MiddlewareControls from '../../../src/app/components/plugins/MiddlewareControls';

jest.mock('../../../src/app/utils/fetchApi');

describe('MiddlewareControls component', () => {
  it('renders', () => {
    const props = {
      url: 'http://localhost:8080',
    };
    const wrapper = shallow(React.createElement(MiddlewareControls, props));
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
