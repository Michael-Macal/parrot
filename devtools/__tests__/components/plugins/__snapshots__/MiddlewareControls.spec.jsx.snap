// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MiddlewareControls component renders 1`] = `
<CircularProgress
  max={100}
  min={0}
  mode="indeterminate"
  size={40}
  thickness={3.5}
  value={0}
/>
`;
